File: wpimath/src/main/native/cpp/autodiff/Variable.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/autodiff/Variable.h"

#include <cmath>
#include <tuple>
#include <vector>

#include <wpi/SymbolExports.h>

namespace frc::autodiff {

Variable::Variable(double value)
    : expr{wpi::MakeIntrusiveShared<Expression>(value)} {}

Variable::Variable(int value)
    : expr{wpi::MakeIntrusiveShared<Expression>(value)} {}

Variable::Variable(wpi::IntrusiveSharedPtr<Expression> expr)
    : expr{std::move(expr)} {}

Variable& Variable::operator=(double value) {
  if (expr == nullptr) {
    expr = wpi::MakeIntrusiveShared<Expression>(value);
  } else {
    GetExpression().value = value;
  }
  return *this;
}

Variable& Variable::operator=(int value) {
  if (expr == nullptr) {
    expr = wpi::MakeIntrusiveShared<Expression>(value);
  } else {
    GetExpression().value = value;
  }
  return *this;
}

WPILIB_DLLEXPORT Variable operator*(double lhs, const Variable& rhs) {
  return Variable{lhs} * rhs;
}

WPILIB_DLLEXPORT Variable operator*(const Variable& lhs, double rhs) {
  return lhs * Variable{rhs};
}

WPILIB_DLLEXPORT Variable operator*(const Variable& lhs, const Variable& rhs) {
  return Variable{lhs.expr * rhs.expr};
}

Variable& Variable::operator*=(double rhs) {
  *this = *this * rhs;
  return *this;
}

Variable& Variable::operator*=(const Variable& rhs) {
  *this = *this * rhs;
  return *this;
}

WPILIB_DLLEXPORT Variable operator/(double lhs, const Variable& rhs) {
  return Variable{lhs} / rhs;
}

WPILIB_DLLEXPORT Variable operator/(const Variable& lhs, double rhs) {
  return lhs / Variable{rhs};
}

WPILIB_DLLEXPORT Variable operator/(const Variable& lhs, const Variable& rhs) {
  return Variable{lhs.expr / rhs.expr};
}

Variable& Variable::operator/=(double rhs) {
  *this = *this / rhs;
  return *this;
}

Variable& Variable::operator/=(const Variable& rhs) {
  *this = *this / rhs;
  return *this;
}

WPILIB_DLLEXPORT Variable operator+(double lhs, const Variable& rhs) {
  return Variable{lhs} + rhs;
}

WPILIB_DLLEXPORT Variable operator+(const Variable& lhs, double rhs) {
  return lhs + Variable{rhs};
}

WPILIB_DLLEXPORT Variable operator+(const Variable& lhs, const Variable& rhs) {
  return Variable{lhs.expr + rhs.expr};
}

Variable& Variable::operator+=(double rhs) {
  *this = *this + rhs;
  return *this;
}

Variable& Variable::operator+=(const Variable& rhs) {
  *this = *this + rhs;
  return *this;
}

WPILIB_DLLEXPORT Variable operator-(double lhs, const Variable& rhs) {
  return Variable{lhs} - rhs;
}

WPILIB_DLLEXPORT Variable operator-(const Variable& lhs, double rhs) {
  return lhs - Variable{rhs};
}

WPILIB_DLLEXPORT Variable operator-(const Variable& lhs, const Variable& rhs) {
  return Variable{lhs.expr - rhs.expr};
}

Variable& Variable::operator-=(double rhs) {
  *this = *this - rhs;
  return *this;
}

Variable& Variable::operator-=(const Variable& rhs) {
  *this = *this - rhs;
  return *this;
}

WPILIB_DLLEXPORT Variable operator-(const Variable& lhs) {
  return Variable{-lhs.expr};
}

WPILIB_DLLEXPORT Variable operator+(const Variable& lhs) {
  return Variable{+lhs.expr};
}

WPILIB_DLLEXPORT bool operator==(double lhs, const Variable& rhs) {
  return std::abs(lhs - rhs.Value()) < 1e-9;
}

WPILIB_DLLEXPORT bool operator==(const Variable& lhs, double rhs) {
  return std::abs(lhs.Value() - rhs) < 1e-9;
}

WPILIB_DLLEXPORT bool operator==(const Variable& lhs, const Variable& rhs) {
  return std::abs(lhs.Value() - rhs.Value()) < 1e-9;
}

WPILIB_DLLEXPORT bool operator!=(double lhs, const Variable& rhs) {
  return lhs != rhs.Value();
}

WPILIB_DLLEXPORT bool operator!=(const Variable& lhs, double rhs) {
  return lhs.Value() != rhs;
}

WPILIB_DLLEXPORT bool operator!=(const Variable& lhs, const Variable& rhs) {
  return lhs.Value() != rhs.Value();
}

WPILIB_DLLEXPORT bool operator<(double lhs, const Variable& rhs) {
  return lhs < rhs.Value();
}

WPILIB_DLLEXPORT bool operator<(const Variable& lhs, double rhs) {
  return lhs.Value() < rhs;
}

WPILIB_DLLEXPORT bool operator<(const Variable& lhs, const Variable& rhs) {
  return lhs.Value() < rhs.Value();
}

WPILIB_DLLEXPORT bool operator>(double lhs, const Variable& rhs) {
  return lhs > rhs.Value();
}

WPILIB_DLLEXPORT bool operator>(const Variable& lhs, double rhs) {
  return lhs.Value() > rhs;
}

WPILIB_DLLEXPORT bool operator>(const Variable& lhs, const Variable& rhs) {
  return lhs.Value() > rhs.Value();
}

WPILIB_DLLEXPORT bool operator<=(double lhs, const Variable& rhs) {
  return lhs <= rhs.Value();
}

WPILIB_DLLEXPORT bool operator<=(const Variable& lhs, double rhs) {
  return lhs.Value() <= rhs;
}

WPILIB_DLLEXPORT bool operator<=(const Variable& lhs, const Variable& rhs) {
  return lhs.Value() <= rhs.Value();
}

WPILIB_DLLEXPORT bool operator>=(double lhs, const Variable& rhs) {
  return lhs >= rhs.Value();
}

WPILIB_DLLEXPORT bool operator>=(const Variable& lhs, double rhs) {
  return lhs.Value() >= rhs;
}

WPILIB_DLLEXPORT bool operator>=(const Variable& lhs, const Variable& rhs) {
  return lhs.Value() >= rhs.Value();
}

double Variable::Value() const {
  if (expr == nullptr) {
    return 0.0;
  } else {
    return expr->value;
  }
}

void Variable::Update() {
  expr->Update();
}

const Expression& Variable::GetExpression() const {
  return *expr;
}

Expression& Variable::GetExpression() {
  return *expr;
}

Variable abs(double x) {
  return autodiff::abs(Variable{x});
}

Variable abs(const Variable& x) {
  return Variable{abs(x.expr)};
}

Variable acos(double x) {
  return autodiff::acos(Variable{x});
}

Variable acos(const Variable& x) {
  return Variable{acos(x.expr)};
}

Variable asin(double x) {
  return autodiff::asin(Variable{x});
}

Variable asin(const Variable& x) {
  return Variable{asin(x.expr)};
}

Variable atan(double x) {
  return autodiff::atan(Variable{x});
}

Variable atan(const Variable& x) {
  return Variable{atan(x.expr)};
}

Variable atan2(double y, const Variable& x) {
  return autodiff::atan2(Variable{y}, x);
}

Variable atan2(const Variable& y, double x) {
  return autodiff::atan2(y, Variable{x});
}

Variable atan2(const Variable& y, const Variable& x) {
  return Variable{atan2(y.expr, x.expr)};
}

Variable cos(double x) {
  return autodiff::cos(Variable{x});
}

Variable cos(const Variable& x) {
  return Variable{cos(x.expr)};
}

Variable cosh(double x) {
  return autodiff::cosh(Variable{x});
}

Variable cosh(const Variable& x) {
  return Variable{cosh(x.expr)};
}

Variable erf(double x) {
  return autodiff::erf(Variable{x});
}

Variable erf(const Variable& x) {
  return Variable{erf(x.expr)};
}

Variable exp(double x) {
  return autodiff::exp(Variable{x});
}

Variable exp(const Variable& x) {
  return Variable{exp(x.expr)};
}

Variable hypot(double x, const Variable& y) {
  return autodiff::hypot(Variable{x}, y);
}

Variable hypot(const Variable& x, double y) {
  return autodiff::hypot(x, Variable{y});
}

Variable hypot(const Variable& x, const Variable& y) {
  return Variable{hypot(x.expr, y.expr)};
}

Variable log(double x) {
  return autodiff::log(Variable{x});
}

Variable log(const Variable& x) {
  return Variable{log(x.expr)};
}

Variable log10(double x) {
  return autodiff::log10(Variable{x});
}

Variable log10(const Variable& x) {
  return Variable{log10(x.expr)};
}

Variable pow(double base, const Variable& power) {
  return autodiff::pow(Variable{base}, power);
}

Variable pow(const Variable& base, double power) {
  return autodiff::pow(base, Variable{power});
}

Variable pow(const Variable& base, const Variable& power) {
  return Variable{pow(base.expr, power.expr)};
}

Variable sin(double x) {
  return autodiff::sin(Variable{x});
}

Variable sin(const Variable& x) {
  return Variable{sin(x.expr)};
}

Variable sinh(double x) {
  return autodiff::sinh(Variable{x});
}

Variable sinh(const Variable& x) {
  return Variable{sinh(x.expr)};
}

Variable sqrt(double x) {
  return autodiff::sqrt(Variable{x});
}

Variable sqrt(const Variable& x) {
  return Variable{sqrt(x.expr)};
}

Variable tan(double x) {
  return autodiff::tan(Variable{x});
}

Variable tan(const Variable& x) {
  return Variable{tan(x.expr)};
}

Variable tanh(double x) {
  return autodiff::tanh(Variable{x});
}

Variable tanh(const Variable& x) {
  return Variable{tanh(x.expr)};
}

}  // namespace frc::autodiff

File: wpimath/src/main/java/edu/wpi/first/math/autodiff/Variable.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.autodiff;

/** An autodiff variable pointing to an expression node. */
public class Variable implements AutoCloseable {
  private long m_handle;

  /** Constructs a linear Variable with a value of zero. */
  public Variable() {}

  /**
   * Constructs a Variable from a floating point type.
   *
   * @param value The value of the Variable.
   */
  public Variable(double value) {
    m_handle = VariableJNI.createDouble(value);
  }

  /**
   * Constructs a Variable from an integral type.
   *
   * @param value The value of the Variable.
   */
  public Variable(int value) {
    m_handle = VariableJNI.createInt(value);
  }

  @Override
  public void close() {
    VariableJNI.destroy(m_handle);
  }

  /**
   * Sets Variable's internal value.
   *
   * @param value The value of the Variable.
   */
  public void setValue(double value) {
    VariableJNI.setValue(m_handle, value);
  }

  /**
   * Variable-Variable multiplication operator.
   *
   * @param rhs Operator right-hand side.
   * @return Result of multiplication.
   */
  public Variable times(Variable rhs) {
    return fromHandle(VariableJNI.times(m_handle, rhs.getHandle()));
  }

  /**
   * Variable-Variable division operator.
   *
   * @param rhs Operator right-hand side.
   * @return Result of division.
   */
  public Variable div(Variable rhs) {
    return fromHandle(VariableJNI.div(m_handle, rhs.getHandle()));
  }

  /**
   * Variable-Variable addition operator.
   *
   * @param rhs Operator right-hand side.
   * @return Result of addition.
   */
  public Variable plus(Variable rhs) {
    return fromHandle(VariableJNI.plus(m_handle, rhs.getHandle()));
  }

  /**
   * Variable-Variable subtraction operator.
   *
   * @param rhs Operator right-hand side.
   * @return Result of subtraction.
   */
  public Variable minus(Variable rhs) {
    return fromHandle(VariableJNI.minus(m_handle, rhs.getHandle()));
  }

  /**
   * Variable-Variable subtraction operator.
   *
   * @param rhs Operator right-hand side.
   * @return Result of subtraction.
   */
  public Variable minus(double rhs) {
    return fromHandle(VariableJNI.minus(m_handle, new Variable(rhs).getHandle()));
  }

  /** Unary minus operator. */
  public Variable unaryMinus() {
    return fromHandle(VariableJNI.unaryMinus(m_handle));
  }

  /**
   * Returns the value of this variable.
   *
   * @return The value of this variable.
   */
  public double value() {
    return VariableJNI.value(m_handle);
  }

  /**
   * Returns the type of this expression (constant, linear, quadratic, or nonlinear).
   *
   * @return The type of this expression.
   */
  public ExpressionType type() {
    return VariableJNI.type(m_handle);
  }

  /**
   * Returns internal handle.
   *
   * @return Internal handle.
   */
  public long getHandle() {
    return m_handle;
  }

  /**
   * Constructs a Variable from an internal handle.
   *
   * <p>This should be used by JNI only.
   *
   * @param handle Variable handle.
   */
  public static Variable fromHandle(long handle) {
    var result = new Variable();
    result.m_handle = handle;
    return result;
  }

  /**
   * std::abs() for Variables.
   *
   * @param x The argument.
   */
  public static Variable abs(Variable x) {
    return fromHandle(VariableJNI.abs(x.getHandle()));
  }

  /**
   * std::acos() for Variables.
   *
   * @param x The argument.
   */
  public static Variable acos(Variable x) {
    return fromHandle(VariableJNI.acos(x.getHandle()));
  }

  /**
   * std::asin() for Variables.
   *
   * @param x The argument.
   */
  public static Variable asin(Variable x) {
    return fromHandle(VariableJNI.asin(x.getHandle()));
  }

  /**
   * std::atan() for Variables.
   *
   * @param x The argument.
   */
  public static Variable atan(Variable x) {
    return fromHandle(VariableJNI.atan(x.getHandle()));
  }

  /**
   * std::atan2() for Variables.
   *
   * @param y The y argument.
   * @param x The x argument.
   */
  public static Variable atan2(Variable y, Variable x) {
    return fromHandle(VariableJNI.atan2(y.getHandle(), x.getHandle()));
  }

  /**
   * std::cos() for Variables.
   *
   * @param x The argument.
   */
  public static Variable cos(Variable x) {
    return fromHandle(VariableJNI.cos(x.getHandle()));
  }

  /**
   * std::cosh() for Variables.
   *
   * @param x The argument.
   */
  public static Variable cosh(Variable x) {
    return fromHandle(VariableJNI.cosh(x.getHandle()));
  }

  /**
   * std::erf() for Variables.
   *
   * @param x The argument.
   */
  public static Variable erf(Variable x) {
    return fromHandle(VariableJNI.erf(x.getHandle()));
  }

  /**
   * std::exp() for Variables.
   *
   * @param x The argument.
   */
  public static Variable exp(Variable x) {
    return fromHandle(VariableJNI.exp(x.getHandle()));
  }

  /**
   * std::hypot() for Variables.
   *
   * @param x The x argument.
   * @param y The y argument.
   */
  public static Variable hypot(Variable x, Variable y) {
    return fromHandle(VariableJNI.hypot(x.getHandle(), y.getHandle()));
  }

  /**
   * std::pow() for Variables.
   *
   * @param base The base.
   * @param power The power.
   */
  public static Variable pow(Variable base, Variable power) {
    return fromHandle(VariableJNI.pow(base.getHandle(), power.getHandle()));
  }

  /**
   * std::pow() for Variables.
   *
   * @param base The base.
   * @param power The power.
   */
  public static Variable pow(Variable base, int power) {
    return fromHandle(VariableJNI.pow(base.getHandle(), new Variable(power).getHandle()));
  }

  /**
   * std::log() for Variables.
   *
   * @param x The argument.
   */
  public static Variable log(Variable x) {
    return fromHandle(VariableJNI.log(x.getHandle()));
  }

  /**
   * std::log10() for Variables.
   *
   * @param x The argument.
   */
  public static Variable log10(Variable x) {
    return fromHandle(VariableJNI.log10(x.getHandle()));
  }

  /**
   * sign() for Variables.
   *
   * @param x The argument.
   */
  public static Variable sign(Variable x) {
    return fromHandle(VariableJNI.sign(x.getHandle()));
  }

  /**
   * std::sin() for Variables.
   *
   * @param x The argument.
   */
  public static Variable sin(Variable x) {
    return fromHandle(VariableJNI.sin(x.getHandle()));
  }

  /**
   * std::sinh() for Variables.
   *
   * @param x The argument.
   */
  public static Variable sinh(Variable x) {
    return fromHandle(VariableJNI.sinh(x.getHandle()));
  }

  /**
   * std::sqrt() for Variables.
   *
   * @param x The argument.
   */
  public static Variable sqrt(Variable x) {
    return fromHandle(VariableJNI.sqrt(x.getHandle()));
  }

  /**
   * std::tan() for Variables.
   *
   * @param x The argument.
   */
  public static Variable tan(Variable x) {
    return fromHandle(VariableJNI.tan(x.getHandle()));
  }

  /**
   * std::tanh() for Variables.
   *
   * @param x The argument.
   */
  public static Variable tanh(Variable x) {
    return fromHandle(VariableJNI.tanh(x.getHandle()));
  }

  /**
   * std::hypot() for Variables.
   *
   * @param x The x argument.
   * @param y The y argument.
   * @param z The z argument.
   */
  public static Variable hypot(Variable x, Variable y, Variable z) {
    return sqrt(pow(x, 2).plus(pow(y, 2)).plus(pow(z, 2)));
  }
}
